datasources/src/main/scala/com/normation/plugins/datasources/Serialization.scala

```diff
@@ -60,368 +66,521 @@ object Translate {
+    value: String
+)
 
-  def extractDataSource(id: DataSourceId, json: JValue) = {
-    for {
-      params <- extractDataSourceWrapper(id, json)
-    } yield {
-      params.to
-    }
+@jsonDiscriminator("name") sealed trait FullDataSourceMode
+
+object FullDataSourceMode {
+  @jsonHint(HttpRequestMode.OneRequestByNode.name) case object OneRequestByNode extends FullDataSourceMode
+  @jsonHint(HttpRequestMode.OneRequestAllNodes.name) case class OneRequestAllNodes(
+      @jsonField("path") matchingPath:       String,
+      @jsonField("attribute") nodeAttribute: String
+  ) extends FullDataSourceMode
+
+  implicit val transformerFrom: Transformer[FullDataSourceMode, HttpRequestMode] =
+    Transformer.derive[FullDataSourceMode, HttpRequestMode]
+  implicit val transformerTo:   Transformer[HttpRequestMode, FullDataSourceMode] =
+    Transformer.derive[HttpRequestMode, FullDataSourceMode]
+}
+
+@jsonDiscriminator("name") sealed trait FullDataSourceMissing
+
+object FullDataSourceMissing {
+  @jsonHint(MissingNodeBehavior.Delete.name) final case object Delete                                extends FullDataSourceMissing
+  @jsonHint(MissingNodeBehavior.NoChange.name) final case object NoChange                            extends FullDataSourceMissing
+  @jsonHint(MissingNodeBehavior.DefaultValue.name) final case class DefaultValue(value: ConfigValue) extends FullDataSourceMissing
+
+  implicit val transformerFrom: Transformer[FullDataSourceMissing, MissingNodeBehavior] =
+    Transformer.derive[FullDataSourceMissing, MissingNodeBehavior]
+  implicit val transformerTo:   Transformer[MissingNodeBehavior, FullDataSourceMissing] =
+    Transformer.derive[MissingNodeBehavior, FullDataSourceMissing]
+}
+
+final case class FullDataSourceRunParameters(
+    onGeneration: Boolean,
+    onNewNode:    Boolean,
+    schedule:     FullDataSourceSchedule
+)
+
+object FullDataSourceRunParameters {
+  implicit val transformerFrom: Transformer[FullDataSourceRunParameters, DataSourceRunParameters] =
+    Transformer.derive[FullDataSourceRunParameters, DataSourceRunParameters]
+  implicit val transformerTo:   Transformer[DataSourceRunParameters, FullDataSourceRunParameters] =
+    Transformer.derive[DataSourceRunParameters, FullDataSourceRunParameters]
+}
+
+@jsonDiscriminator("type") sealed trait FullDataSourceSchedule
+
+object FullDataSourceSchedule {
+  @jsonHint("scheduled") final case class Scheduled(
+      duration: Duration
+  ) extends FullDataSourceSchedule
+  @jsonHint("notscheduled") final case class NoSchedule(
+      @jsonField("duration") savedDuration: Duration
+  ) extends FullDataSourceSchedule
+
+  implicit val transformerFrom: Transformer[FullDataSourceSchedule, DataSourceSchedule] =
+    Transformer.derive[FullDataSourceSchedule, DataSourceSchedule]
+
+  implicit val transformerTo: Transformer[DataSourceSchedule, FullDataSourceSchedule] =
+    Transformer.derive[DataSourceSchedule, FullDataSourceSchedule]
+}
+
+trait DataSourceJsonCodec {
```

Comment:
I'm not sure having that as a trait IF extening traits neet to use the same encoders are a great idea. 
Depending of the order of init, you can get error like that: 
```
java.lang.ExceptionInInitializerError
	at com.normation.plugins.datasources.DataSourceExtractor$PatchDataSourceType$.<clinit>(Serialization.scala:442)
	at com.normation.plugins.datasources.DataSourceJsonCodec.$init$(Serialization.scala:366)
	at com.normation.plugins.datasources.DataSourceJdbcRepository$.<clinit>(Repository.scala:483)
	at com.normation.plugins.datasources.RepositoryTest.$anonfun$new$2(RepositoryTest.scala:92)
	at org.specs2.specification.dsl.mutable.EffectBlocks.tryBlock(EffectBlocks.scala:131)
	at org.specs2.specification.dsl.mutable.EffectBlocks.$anonfun$nestBlock$2(EffectBlocks.scala:110)
        ....
	at com.normation.plugins.datasources.DataSourceExtractor$PatchDataSourceType$.decoder(Serialization.scala:443)
	at com.normation.plugins.datasources.DataSourceJsonCodec.$init$(Serialization.scala:366)
	at com.normation.plugins.datasources.DataSourceExtractor$.<clinit>(Serialization.scala:387)
	... 43 more
```

Here, we have a call to 
```
      check(DataSourceJdbcRepository.insertDataSourceSQL(fakeDataSource.id, fakeDataSource.transformInto[FullDataSource]))
```
which extends `DataSourceJsonCodec`. So we init parent (trait) parameters, ie all the `implicit val` in order. 
We reach the last one, `NewDataSource`, which has a `PatchDataSourceHttpType`. 

So we look for the codec, which is provided by `object PatchDataSourceType` in `object DataSourceExtractor extends DataSourceJsonCodec`. 
So, same story, parent `DataSourceJsonCodec` are all init, that's an other object so we pay the init cost again, and we reach the point where we reach `NewDataSource` again. 

And here, it tries to init its own `PatchDataSourceHttpType`, but (I think) the init lock for that object is hold by `DataSourceJdbcRepository`, so the field is null. 

Basically, when you use trait, you should be sure that: 
- either you have `val`s but with ONLY ONE direct, linear inheritage line, ie:  ` trait A { ... } ; trait B extends A { ... } ; object C extends B` and that the init order is linear from A to B to C. You can also have several linear hierarchy converging in the last object which "ties the knot". 
- or you use `lazy val` everywhere. Even then, you're not garanted to get all the lock correctly, because scala object lazy init and lock are not proved to work in all case (and they are not, we can trigger the problem easily like here, or here: https://issues.rudder.io/issues/22645 (we has a deadlock, ie the opposite problem), or in scalac bugtracker). So you need to put your object init in a `def`, and it's not very elegant. 

So, I would say for here: 
- don't use trait to hold common data. They are not done to do that, and it leads to surprising init conemdrum, and it duplicate all the data structures (which also means that you will avoid to get jvm nice optim on codec, because depending of the path, it won't be the same object's codec for `DataSourceJsonCodec`.
- if you want, create just one linear hierarchy for Codec, and nothing else. 
- then import the last object which ties the knot where codec are needed, don't extends it.

Reply:
Thanks for the super clear comment and suggestions ! I like having trait to have that 'hierarchy' and a namespace of codec definitions, but I didn't think about those issues...
I removed the trait as there was no need for it here, I changed it to an `object` :+1: